CP-46179 Deterministic UUID for Back-Up VDI

lindig opened the pull request:
When creating a backup VDI on an SR we want to derive the VDI's UUID
   from the SR. This way we can't be later tricked into accepting a
   different VDI as a backup.
    
   Implement a hash that derives the VDI UUID and pass this UUID to SM to
   be used.
    
   Currently the creation of this VDI in Xapi is detected based on its
   name-label: "Pool Metadata Backup". The SM stack usually creates a new
   random UUID but will use a given UUID passed as "vdi_uuid" for its
   "vdi_create" command. The difficulty in the implementation is that
   "vdi_uuid" is used in other commands as well and we have to make sure to
   set it only in the intended context. It is not straight forward to pass
   the UUID from Xapi_vdi.create down to Sm.

This will require an updated xapi.spec to package the changes.

robhoes commented on ocaml/libs/uuid/uuidx.ml:
```diff
@@ -83,3 +83,15 @@ let string_of_uuid = to_string
 
 let int_array_of_uuid = to_int_array
+
+module XenServer = struct
```
This name is not really appropriate, as this is Xapi Project code, not exclusive to XenServer. Something related to the actual use case may be better?

robhoes commented on ocaml/xapi/storage_smapiv1.ml:
```diff
@@ -691,19 +691,33 @@ module SMAPIv1 : Server_impl = struct
+              let vdi_uuid =
+                match vdi_info.uuid with
+                | Some uuid
+                  when uuid = Uuidx.(XenServer.hash sr_uuid |> to_string) ->
```
This is just a sanity check to be extra sure that this uuid is only used in the intended use case?

robhoes commented on scripts/xe-backup-metadata:
```diff
@@ -134,11 +156,25 @@ function cleanup {
+# if we can't validate the UUID of the VDI, prompt the user
+if [ -n "${vdi_uuid}" ]; then
+    if ! validate_vdi_uuid "${sr_uuid}" "${vdi_uuid}"; then
+        local version="8.4"
+        echo "Backup VDI $vdi was created prior to XenServer $version."
```
We should avoid the version number and name XenServer here.

robhoes replied:
This is the case where an existing backup VDI was found, which will be reused? I think the prompt should make that clearer.

robhoes reviewed (approved):
A couple of small comments, but otherwise I think this is fine.

edwintorok commented on ocaml/xapi/xapi_vdi.ml:
```diff
@@ -625,13 +625,27 @@ let create ~__context ~name_label ~name_description ~sR ~virtual_size ~_type
+     Backup *)
+  let uuid_ =
+    match (_type, name_label) with
+    | `user, "Pool Metadata Backup" ->
```
Will the VDI type and name label still be preserved in the scenario we want to use this for, i.e. recovery of a lost XAPI database?
I think all this info would've been stored in the xapi database itself, the only thing you can know about the VDI is its UUID, no other metadata would survive a loss of the XAPI database.

edwintorok replied:
Would be good to have a test for this (obviously not here, but in XenRT) that tries to exercise the scenario where this backup/recovery is meant to be used and check that you can still recover starting from a fresh install, and attaching an SR.

edwintorok commented on scripts/xe-backup-metadata:
```diff
@@ -48,6 +48,28 @@ function usage {
     exit 1
 }
```
this script is missing a generic `set -eu`: errors or uninitialised variables would not cause the script to stop, but it'll keep going, potentially executing the wrong things.

edwintorok commented on scripts/xe-backup-metadata:
```diff
@@ -134,11 +156,25 @@ function cleanup {
    fi
```
We should run `ShellCheck` in the CI, there are about 66 warnings reported on this file, which means that about 20% of this script is slightly wrong (pre-existing problem, not necessarily with the changes here). I'll attempt to paste the relevant output from shellcheck as comments.

edwintorok commented on scripts/xe-backup-metadata:
```diff
@@ -134,11 +156,25 @@ function cleanup {
+    fi
+fi
+
 echo Using SR: ${sr_name}
```
In xe-backup-metadata line 171:
echo Using SR: ${sr_name}
               ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
echo Using SR: "${sr_name}"

edwintorok commented on scripts/xe-backup-metadata:
```diff
@@ -134,11 +156,25 @@ function cleanup {
-    vdi_uuid=$(${XE} vdi-create virtual-size=500MiB sr-uuid=${sr_uuid} type=user name-label="Pool Metadata Backup")
+    label="Pool Metadata Backup"
+    # the label must match what xapi_vdi.ml is using for backup VDIs
+    vdi_uuid=$(${XE} vdi-create virtual-size=500MiB sr-uuid=${sr_uuid} type=user name-label="${label}")
```
In xe-backup-metadata line 177:
    vdi_uuid=$(${XE} vdi-create virtual-size=500MiB sr-uuid=${sr_uuid} type=user name-label="${label}")
                                                            ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
    vdi_uuid=$(${XE} vdi-create virtual-size=500MiB sr-uuid="${sr_uuid}" type=user name-label="${label}")

edwintorok commented on scripts/xe-backup-metadata:
```diff
@@ -134,11 +156,25 @@ function cleanup {
+    # the label must match what xapi_vdi.ml is using for backup VDIs
+    vdi_uuid=$(${XE} vdi-create virtual-size=500MiB sr-uuid=${sr_uuid} type=user name-label="${label}")
     init_fs=1
     if [ $? -ne 0 ]; then
```
In xe-backup-metadata line 179:
    if [ $? -ne 0 ]; then
         ^-- SC2181 (style): Check exit code directly with e.g. 'if ! mycmd;', not indirectly with $?.

edwintorok commented on scripts/xe-restore-metadata:
```diff
@@ -100,6 +107,13 @@ test_sr "${sr_uuid}"
 
 sr_name=$(${XE} sr-param-get uuid=${sr_uuid} param-name=name-label)
```
In xe-restore-metadata line 108:
sr_name=$(${XE} sr-param-get uuid=${sr_uuid} param-name=name-label)
^-----^ SC2034 (warning): sr_name appears unused. Verify use (or export if used externally).
                                  ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
sr_name=$(${XE} sr-param-get uuid="${sr_uuid}" param-name=name-label)

edwintorok commented on scripts/xe-restore-metadata:
```diff
@@ -100,6 +107,13 @@ test_sr "${sr_uuid}"
+
 # get a list of all VDIs if an override has not been provided on the cmd line
 if [ -z "${vdis}" ]; then
   vdis=$(${XE} vdi-list params=uuid sr-uuid=${sr_uuid} read-only=false --minimal)
```
In xe-restore-metadata line 119:
  vdis=$(${XE} vdi-list params=uuid sr-uuid=${sr_uuid} read-only=false --minimal)
                                            ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
  vdis=$(${XE} vdi-list params=uuid sr-uuid="${sr_uuid}" read-only=false --minimal)

edwintorok reviewed (commented):
Github doesn't let me add comments on lines that haven't changed.

Most of them are probably just style issues (a lot of the strings here are uuids, so we know they won't contain spaces), but these look like important ones to fix:
```
In xe-backup-metadata line 154:
   if [ ${fs_uninitialised} -eq 1 -a -n "${vdi_uuid}" ] ; then
                                  ^-- SC2166 (warning): Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.
```

This would result in a runtime error (sounds like we'd need a test to trigger this codepath, because this should result in an immediate runtime error "bash: local: can only be used in a function")
```
In xe-backup-metadata line 162:
        local version="8.4"
        ^---^ SC2168 (error): 'local' is only valid in functions.

```

```
In xe-backup-metadata line 264:
    echo "Running out of space, you can use "-d" option to attach VDI and free more space, exit now."
                                             ^-- SC2140 (warning): Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"?

```

This appears to be somewhat problematic, can this result in a shell injection?
```
In xe-backup-metadata line 286:
  todelete=$(cd ${mnt}/metadata && ls -1 |sort -n | head -n -${history_kept} | xargs echo)
                ^----^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                   ^---^ SC2012 (info): Use find instead of ls to better handle non-alphanumeric filenames.
                                                             ^-------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
Did you mean:
  todelete=$(cd "${mnt}"/metadata && ls -1 |sort -n | head -n -"${history_kept}" | xargs echo)

```

Similar warnings in the other script:
```
In xe-restore-metadata line 254:
    chosen_metadata_dir=$(ls | sort -n | tail -1)
                          ^-- SC2012 (info): Use find instead of ls to better handle non-alphanumeric filenames.

```

There are a lot of warning about the use of unquoted variables, too many to list here, just one that looks interesting is this one, not sure whether this can be attacker controlled, best to quote it to be safe:
```
In xe-restore-metadata line 278:
if [ ! -d ${full_dir} ]; then
          ^---------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
if [ ! -d "${full_dir}" ]; then


In xe-restore-metadata line 284:
${debug} echo Selected: ${full_dir}
                        ^---------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
${debug} echo Selected: "${full_dir}"

```

```
In xe-restore-metadata line 293:
   echo "Please type in "yes" and <enter> to continue."
                         ^-^ SC2140 (warning): Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"?


```

We need to be careful with other potentially attacker controlled paths:
```
In xe-restore-metadata line 322:
   echo xe vm-import filename=${meta} sr-uuid=${sr_uuid} --metadata --preserve${force_flag}${dry_run_flag}
                              ^-----^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                              ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                                              ^-----------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                                                           ^-------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
   echo xe vm-import filename="${meta}" sr-uuid="${sr_uuid}" --metadata --preserve"${force_flag}""${dry_run_flag}"
```

Some more:
```
In xe-restore-metadata line 338:
if [ -e ${smmeta_file} ]; then
        ^------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
if [ -e "${smmeta_file}" ]; then


In xe-restore-metadata line 340:
        echo ${cmd}
             ^----^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
        echo "${cmd}"
```

```

lindig commented:
After a long break:

* I've addressed comments
* Improved the shell quoting but tried to not touch too much code because it is risky
* This code passed some manual testing and some XenRT testing
* I've added a `-y` (yes) flag to supporting scripting after this code asks for confirmation when working with existing backup VDIs.
* To ship this code we need to update `xapi.spec`; it is currently not shipping.

edwintorok commented:
I looked at the shellcheck output again and other than stylistic issues I think there is this one bug remaining:
```
In xe-backup-metadata line 275:
  num_found=$(find "${mnt}" -name \*.db\.* | wc -l)
                                  ^------^ SC2061 (warning): Quote the parameter to -name so the shell won't interpret it.
```

The escaping there looks incorrect: the last '*' will be interpreted and expanded by the shell, not by 'find'. It may happen to work if you don't have any failed called '*.db.' in the current dir, and fail if you do (because then the shell expands it based on the current dir, not ${mnt}'.
I think the final `*` needs to be escaped, i.e. `find "${mnt}" -name \*.db\.\* | wc -l`

And there are also a couple of https://www.shellcheck.net/wiki/SC2166, but I can't tell whether we are affected or not. The script is using a syntax that has been deprecated by POSIX because it doesn't work if you have dashes in the arguments, and we do have dashes in the arguments. We can leave those as is.

lindig commented:
I am not sure what the command tries to do. In particular, why is the `.` escaped? I assume this is a glob and not a regex. That is why I did not change it.

edwintorok commented:
Maybe it was a typo, they meant to escape the '*' but escaped the '.' by accident.

lindig commented:
I agree. I think `find "${mnt}" -name '*.db.*'` would do the right thing. The files the code is looking are:
```
a45301a0-3eaf-8533-9281-9057cc39c779.db
a45301a0-3eaf-8533-9281-9057cc39c779.db.1
a45301a0-3eaf-8533-9281-9057cc39c779.db.2
```

edwintorok commented:
Yes, using apostrophes, or escaping the * would both fix the problem:
```
# echo \*.db\.*
*.db.* *.db.A
# echo \*.db.\*
*.db.*
# echo '*.db.*'
*.db.*
```
